src/Microsoft.ML.Data/Utilities/TypeParsingUtils.cs

```diff
@@ -116,19 +106,18 @@ private bool TryParse(string str)
         {
             Contracts.AssertValue(str);
 
+            ulong min;
             int ich = str.IndexOf('-');
             if (ich < 0)
             {
-                if (!ulong.TryParse(str, out Min))
+                if (!ulong.TryParse(str, out min))
                     return false;
-                Contiguous = false;
                 return true;
             }
 
-            if (!ulong.TryParse(str.Substring(0, ich), out Min))
+            if (!ulong.TryParse(str.Substring(0, ich), out min))
```

Comment:
>min [](start = 59, length = 3)

What if `min` is non-zero, since we no longer support that? What then?

Reply:
We need to think more carefully about what the format here should be. Just leaving it unchanged is clearly wrong (since it is meant to support some concepts that no longer exist), but what woudl be right, I am less certain about. Maybe @yaeldekel has some thoughts here.

---
In reply to: [249262458](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249262458) [](ancestors = 249262458)

Reply:
I initially thought that this would come with a subsequent PR. But I think it can be easily done as part of this PR. 

Currently the format looks like: "TYPE[MIN-COUNT]". I think the format should change to something that looks like "TYPE[COUNT]", so something like this: U4[100].

This removes the Min, and we can also enforce the Count to be required (so that "*" is not acceptable for Count).
It should be pretty easy to do this change. 

---
In reply to: [249262473](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249262473) [](ancestors = 249262473,249262458)

Reply:
Maybe parts of it can be done in a separate PR, but I think those parts of the syntax that are meant to support features that no longer exist should be made to throw now, at least. So, `0-5` is fine, but `2-5` or `0-*` or `0-` is not for `KeyRange`, would be the idea, since where used, those things, even though we never used them, had a definite meaning and we need to be clear that's no longer supported.

...Not that anything other than our tests really used it, but, you know, still. 😄

Reply:
Also when you say "removes the Min", I think you also mean changing the format so that instead of `U4[0-10]` or something it would be just `U4[10]`. While this is fine, I think it is important that at least for the short term we continue to support `U4[0-10]` as an *input* format at least, so that older text files from prior versions should be readable, for at least some versions, until we determine that support is no longer necessary.

---
In reply to: [249625047](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249625047) [](ancestors = 249625047)

Reply:
From discussing with Tom and Yael, we decided to do the following:

Change the format to `U4[100]`, or `U4[]` and maintain backwards compatibility with `U4[0-100]` and `U4[0-*]` throwing if `Min`  is not 0. 

---
In reply to: [249626204](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249626204) [](ancestors = 249626204,249625047)